### ISSUE TITLE
Password-protected messages compliance

### ISSUE THREAD

**Issue by sosnovsky** [FlowCrypt/flowcrypt-android#2905]
To ensure password-protected messages comply with security rules, client apps must check the subjects of composed password-protected email messages for specified strings. If any configured terms are found, the application should display an error message indicating that the password encryption method is incompatible with the composed message.

Client configuration will have 2 new optional properties (subject check should be performed only if both properties are set):
- `disallow_password_messages_for_terms` (_string[]_) - specifies an array of strings to check against the subject of the composed password-protected message. If any string in this array is found in the subject, an error alert must be displayed.
- `disallow_password_messages_error_text` (_string_) - defines the text to be displayed in the error alert. May contain URLs which should be auto-detected and be clickable in the displayed alert (for example, `Password-protected messages are disabled, please check https://flowcrypt.com`)

related https://github.com/FlowCrypt/flowcrypt-browser/issues/5878

**Comment by sosnovsky** [comment]
Hello @DenBond7, can you please pause your work on thread functionality and implement this feature, as it's a customer request which needs to be released sooner.
And after implementing this one, let's make a new release and continue working on threads feature, as it'll require a thoroughly testing before publishing to play store. Thanks!

**Comment by DenBond7** [comment]
Sure, I'm switching right now.

**Comment by DenBond7** [comment]
> disallow_password_messages_for_terms (string[]) - specifies an array of strings to check against the subject of the composed password-protected message. If any string in this array is found in the subject, an error alert must be displayed.

@sosnovsky is it a case-sensitive search?

for example, I have the following subject: `Nise subject with Android` and
```json
"disallow_attester_search_for_domains":[
         "Android", //1
         "ANDROID",//2
         "android",//3
         "AnDroid",//4
 ]
```
Which option will be matched? Only `1` or all?

**Comment by sosnovsky** [comment]
Good point! Let's make it's case-insensitive, so all options will match in your example

**Comment by DenBond7** [comment]
I have a few more questions

1) > ...messages for specified strings. If any configured terms are found, the application should display an error message indicating that the password encryption method is incompatible with the composed message.

What about the following case?

subject: `Android developers from Europe and across the globe will converge this year at droidcon in London`

```json
"disallow_attester_search_for_domains":[
         "Droid",
         "ANDROID",
 ]
```
We have the word `droidcon` and one term `Droid`. Is it our case or should the search be word-oriented? Like

subject: `Android developers from Europe and across the globe will converge this year at DROID con in London`

 2) > disallow_password_messages_for_terms (string[]) - specifies an array of strings to check against the subject of the composed password-protected message. If any string in this array is found in the subject, an error alert must be displayed.

Should I expect that `disallow_password_messages_for_terms` will contain only non-empty strings?

Like 
```json
"disallow_attester_search_for_domains":[
         "Droid",
         "ANDROID",
 ]
```

but not like

```json
"disallow_attester_search_for_domains":[
         "",
         "",
 ]
```

**Comment by DenBond7** [comment]
tests
- [ ] case when `disallow_password_messages_for_terms` and `disallow_password_messages_error_text` are not present
- [ ] case when `disallow_password_messages_for_terms` is present, `disallow_password_messages_error_text` is not present
- [ ] case when `disallow_password_messages_for_terms` is not present, `disallow_password_messages_error_text` is present
- [ ] case for a few terms in subject, URL in the error text
```json
"subject":"Android developers from Europe and across the globe will converge this year at droidcon in London",
"disallow_password_messages_for_terms":[
   "Droid",
   "Android",
   "ANDROID",
   "android",
   "AnDroid"
],
"disallow_password_messages_error_text":"Password-protected messages are disabled, please check https://flowcrypt.com",

```

**Comment by sosnovsky** [comment]
> We have the word droidcon and one term Droid. Is it our case or should the search be word-oriented?

Let's just search for the same string, not word-oriented. It's expected to have some specific terms in `disallow_password_messages_for_terms`, which won't be a common words, but some abbreviations for categorizing emails.

> Should I expect that disallow_password_messages_for_terms will contain only non-empty strings?

I think there won't be empty strings, but let's have additional check for non-empty strings, to be sure that configuration is correct.

Thanks for such detailed questions!

**Comment by sosnovsky** [comment]
> We have the word droidcon and one term Droid. Is it our case or should the search be word-oriented? 

I also asked customer about this case, to be sure that it works correctly.

**Comment by sosnovsky** [comment]
I got customer's response:
- search should be case-insensitive
- it should work only for exact matches - if we have `droid` in the list of strings, password-protected messages shouldn't be disabled for subjects with `Android` word

Also they give some examples of possible classification strings - `['ABDU', '[Classification: Data Control: Internal Data Control]']`.

Please let me know if you have any additional questions.

**Comment by ioanmo226** [comment]
Could you explain more about the classification strings mentioned above? I'm not entirely sure how the classification works.

Can `disallow_password_messages_for_terms` also contain a string array? If so, is it separated by `:`?

For example, in the case of `['ABDU', '[Classification: Data Control: Internal Data Control]']`, could you provide possible disallowed subjects?

Would the disallowed subjects be something like `ABDU subject`, `Data Control subject`, or `Internal Data subject`?

Additionally, is the term `Classification` in `[Classification: ...]` static, or can it be replaced with any other word?

**Comment by ioanmo226** [comment]
Also will `disallow_password_messages_for_terms` contain strings with group of words? (like ['droid', 'Android developers'])?
Example subject: `Android developers from Europe and across the globe will converge this year at droidcon in London`

**Comment by sosnovsky** [comment]
> Could you explain more about the classification strings mentioned above? I'm not entirely sure how the classification works.

`[Classification: Data Control: Internal Data Control]` is full classification string, so only email subjects which contain exact `[Classification: Data Control: Internal Data Control]` string will match. We don't need to do any parsing of strings from `disallow_password_messages_for_terms`, just search for full string match.

For this case matches will work this way:
✅ [Classification: Data Control: Internal Data Control] Quarter results
✅ Conference information [Classification: Data Control: Internal Data Control]
❌ Classification: Data Control: Internal Data Control - Tomorrow meeting
❌ Internal Data Control - Finance monitoring

> Also will disallow_password_messages_for_terms contain strings with group of words? (like ['droid', 'Android developers'])?

Yes, multiple words are possible too, like in `[Classification: Data Control: Internal Data Control]` example.
